cloud/ec2_client.go

```diff
@@ -151,34 +145,29 @@ func awsClientDefaultRetryOptions() utility.RetryOptions {
 	}
 }
 
+var configCache map[string]*aws.Config = make(map[string]*aws.Config)
+
 // Create a new aws-sdk-client if one does not exist, otherwise no-op.
 func (c *awsClientImpl) Create(ctx context.Context, region string) error {
 	if region == "" {
 		return errors.New("region must not be empty")
 	}
-	if c.config == nil {
-		c.httpClient = utility.GetHTTPClient()
+
+	if configCache[region] == nil {
```

Comment:
Could we cache the whole config and on use, just change the Region field value? I only took a quick look inside the config and how it handles and it looks like everything is pass by value and the region only affects this property (until we make the client).

Reply:
I'm with you that this suggestion would simplify the code a lot.

I'm worried, though, that the region that's set when we first load the config is persisted in insidious ways. One thing I saw (because I was looking through the SDK's credentials code) was that the region is set in the credentials provider that we use to call STS. This provider is later used to get creds for the derived clients. So if we specify a region when we first load the config and then later change it perhaps the STS calls will go to the first region so the creds won't be valid for calls to other regions.

We could maybe not specify a region when we first make a config and then change it when we make the clients. This would probably work for STS at least since [it has a global endpoint that's not region specific](https://docs.aws.amazon.com/IAM/latest/UserGuide/id_credentials_temp_enable-regions.html#:~:text=By%20default%2C%20the,session%20token%20validity.), though I don't know if there are other things. I'm worried to do this as well: [the docs](https://aws.github.io/aws-sdk-go-v2/docs/configuring-sdk/#specifying-the-aws-region) say to specify it when constructing the config.